shared/src/test/diff/mlscript/RecursiveTypes.mls

```diff
@@ -307,39 +307,58 @@ f { a = 1 }
 //│ ╟── expression of type `int` does not have field 'a'
 //│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
-//│ ╟── but it flows into record literal with expected type `~{a: ?a} | {a: ?b}`
+//│ ╟── but it flows into record literal with expected type `{a: ?a} | ~{a: ?b} | ~{}`
 //│ ║  l.303: 	f { a = 1 }
 //│ ║         	  ^^^^^^^^^
 //│ ╟── Note: constraint arises from field selection:
 //│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
-//│ res: error | int | ({a: 1} | 'a\a & {a: int} as 'a)
+//│ res: error
 
 f error
 //│ res: 'a\a & {a: int} as 'a
 
+
+:e
+rec def ainf = { a = ainf }
+f ainf
+//│ ainf: {a: 'a} as 'a
+//│ ╔══[ERROR] Type mismatch in application:
+//│ ║  l.324: 	f ainf
+//│ ║         	^^^^^^
+//│ ╟── expression of type `int` does not have field 'a'
+//│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║         	                                             ^^^^^
+//│ ╟── but it flows into reference with expected type `({a: ?a} | ~{a: ?b} | ~{})\a`
+//│ ║  l.324: 	f ainf
+//│ ║         	  ^^^^
+//│ ╟── Note: constraint arises from field selection:
+//│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ╙──       	                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+//│ res: error
+
 :e
 rec def infina = 0 with { a = infina }
 f infina
 //│ infina: 0 & {a: 'a} as 'a
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.323: 	f infina
+//│ ║  l.342: 	f infina
 //│ ║         	^^^^^^^^
 //│ ╟── expression of type `int` does not have field 'a'
 //│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
-//│ ╟── but it flows into reference with expected type `(~{a: ?a} | {a: ?b})\a`
-//│ ║  l.323: 	f infina
+//│ ╟── but it flows into reference with expected type `({a: ?a} | ~{a: ?b} | ~{})\a`
+//│ ║  l.342: 	f infina
 //│ ║         	  ^^^^^^
 //│ ╟── Note: constraint arises from field selection:
 //│ ║  l.275: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error | int | (('b | (0 & {a: 'c} as 'c) | 'a)\a & {a: int} as 'a)
 
 
-// FIXME why does type printing show `nothing`?
+// TODO why do we get this weird `'a as 'a`? (Note: it used to show `nothing`.)
```

Comment:
I think this is because empty `Without`s are created in `LhsNf.&`, and they are not simplified away. So it's actually `Without('a, Set.empty) as 'a`. I think simplifying them away in `SimplifyType.transform` fixes them without any side-effects. (_I think..._ I'm not too familiar with the constraining and simplification parts.)

Reply:
Aha, we have a winner 😄 

Good catch!